# Transcriptome module workflows

The transcriptome module is divided in three entrypoints: A single general entrypoint that encompasses the whole analysis, a read alignment and assembly entrypoint and a mikado entrypoint which uses the results of the read alignment. These steps are represented respectively in the files `main.wdl`, `align.wdl` and `mikado.wdl`.

* `main.wdl` - Defines the complete workflow entrypoint which makes use of `align.wdl` and `mikado.wdl`.
* `align.wdl` - Defines the alignment of reads and assembling of transcripts.
* `mikado.wdl` - Defines the evaluation and scoring of transcripts, along with the inclusion of external evidence such as species related proteins.

## Align workflow (`align.wdl`)

Mention all the possible choices for aligners (short, and long) and assemblers (short, and long) along with the filtering options (merge, merge_collapse and filter_gff).

The align workflow is comprised of alignment, assembly and splice junction curation using portcullis for short reads and alignment and assembly of long reads informed by predetermined splice junctions (via portcullis or reference guided).

### Alignment

The alignment step for paired short reads (PSR) and single-end short reads (SSR) allows the user to choose between `hisat2` and `star` splice aware short RNA read aligners. PSR and SSR labels are marked by a user provided strandness which will be used by the aligners and portcullis. 

Besides the choice of aligner, how these alignments are carried out is also a choice for the user. PSR files can be labelled together and be aligned separately or merged before the assembly step, similarly for SSR. If the files are aligned separetely, their assemblies will be joined, otherwise, the alignments will be merged and then assembled.

The alignment step for long reads (LR) is specialised for high quality (HQ) and low quality (LQ) sequencing technologies. Reads are considered high quality if their base accuracy is over 95%(?), this is generally the case for circular consensus Pacbio reads. If available, files containing HQ reads can be provided into one or more labels, similarly for LQ reads. HQ and LQ labels are then aligned based on the choice of aligner `minimap2` or `gmap` using a set of overridable default parameters selected for best performance on HQ and LQ data.

Alignment summary statistics are generated for all alignments under each label for short and long read data.

TODO: Add an example of the statistics and the configuration section.

### Assembly

Short read assemblies are generated by both `stringtie` and `scallop` based on a label, a strandness and a list of one or more alignment files (BAM).

For `stringtie`, when multiple BAM files are provided under a single label, each BAM file will be assembled separately and finally merged into a single assembly. For more details on this, please look at the `merge` parameter of the read configuration. 

For `scallop`, multiple files will be assembled together and the choice as the case for `stringtie` is based on the presence of a single or multiple BAM files under each label.

Assembly summary statistics are generated for each assembler program and each label.

TODO: Add an example of the statistics and the configuration section.

There are four options available for long read assemblies: `filter`, `merge`, `stringtie` and `stringtie_collapse`.

The `filter` option for long read assemblies, generates models directly based alignments which are filtered by quality and coverage. Default quality and coverage parameters can be overriden using the configuration file.

The `merge` option uses `gffread` merge to combine transcripts representing the same region of the genome.

`stringtie` generates assemblies from the long read alignments, whilst `stringtie_collapse` uses the `-R` option simply collapsing alignments and not performing any assembly.

Statistics are generated for each assembly where the process of obtaining this assembly has not been `filter` and a final table of summary statistics is provided to compare results between labels.

TODO: Add an example of the statistics and the configuration section.

### Portcullis

TODO: Explain how portcullis is used.

A diagram would be helpful here.

## Mikado workflow (`mikado.wdl`)

Define in words what this is. Link to the Mikado paper as a reference.

Add a diagram of mikado + ancilliary processes (ORFs + Homology).
